### PR TITLE
common/ceph_context: leak some memory fail to show in valgrind

### PR DESCRIPTION
**Pull request by NitzanMordhai**
GCC 20 memory optimization broke the memory leak for testing.

Fixes:https://tracker.ceph.com/issues/57165
Signed-off-by: Nitzan Mordechai <nmordec@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/common/ceph_context.cc [hunk 1]
```diff
@@ -502,6 +502,7 @@ int CephContext::do_command(std::string_view command, const cmdmap_t& cmdmap,
   }
 }
 
+
```
Comment (tchaikov) [inline src/common/ceph_context.cc:505]: irrelevant change.

File: src/common/ceph_context.h [hunk 2]
```diff
@@ -258,7 +258,13 @@ class CephContext {
...
     return _set_gid_string;
   }
-
+  #pragma GCC push_options
+  #pragma GCC optimize ("O0")
+  void leak_some_memory() {
+    volatile char *foo = new char[1234];
+    (void)foo;
+  }
+  #pragma GCC pop_options
```
Comment (tchaikov) [inline src/common/ceph_context.h:267]: can we put this back to the .cc file? just to improve the compile-time performance?

Comment (NitzanMordhai) [reply src/common/ceph_context.h]: done

File: src/common/ceph_context.h [hunk 3]
```diff
@@ -258,7 +258,7 @@ class CephContext {
   std::string get_set_gid_string() const {
     return _set_gid_string;
   }
-
+  void leak_some_memory();
```
Comment (tchaikov) [inline src/common/ceph_context.h:261]: so there is no need to have this declaration in header file, right? if that's the case, we can define it as a `static` function in .cc without this declaration.

**Review by ljflores — APPROVED** [review]
Approving the concept; +1 to the change @tchaikov requested

File: src/common/ceph_context.cc [hunk 4]
```diff
@@ -502,6 +502,14 @@ int CephContext::do_command(std::string_view command, const cmdmap_t& cmdmap,
   }
 }
 
+#pragma GCC push_options
+#pragma GCC optimize ("O0")
+void leak_some_memory() {
```
Comment (tchaikov) [inline src/common/ceph_context.cc:507]:
```suggestion
static void leak_some_memory() {
```

**Review by tchaikov — APPROVED** [review]

**Review by rzarzynski — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by NitzanMordhai** [comment]
jenkins test make check

**Comment by rzarzynski** [comment]
```
The following tests FAILED:
	 33 - run-rbd-unit-tests-109.sh (Failed)
	 34 - run-rbd-unit-tests-127.sh (Failed)

```

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2022-08-30_00:38:22-rados-wip-kefu-testing-2022-08-29-1815-distro-default-smithi/

failures tracked by https://tracker.ceph.com/issues/57332

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Pull request merged**